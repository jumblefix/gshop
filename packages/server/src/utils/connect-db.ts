// tslint:disable-next-line:no-var-requires
const iconv = require('iconv-lite');
// tslint:disable-next-line:no-var-requires
const encodings = require('iconv-lite/encodings');
iconv.encodings = encodings;

import { ConnectionOptions, createConnection } from 'typeorm';

const {
  DB_HOST,
  DB_PORT,
  DB_USER,
  DB_PASS,
  DB_NAME,
  DB_TYPE,
  DB_NAME_TEST,
  ROOT_USER,
  ROOT_PASS,
  DB_NAME_PRODUCTION
} = process.env;

export interface ConnectionList {
  [key: string]: any;
}

export const c: ConnectionList = {
  development: {
    name: 'development',
    type: DB_TYPE,
    host: DB_HOST,
    port: DB_PORT,
    username: DB_USER,
    password: DB_PASS,
    database: DB_NAME,
    dropSchema: true,
    synchronize: true,
    logging: true,
    entities: ['src/entity/**/*.ts'],
    migrations: ['src/migration/**/*.ts'],
    subscribers: ['src/subscriber/**/*.ts'],
    cli: {
      entitiesDir: 'src/entity',
      migrationsDir: 'src/migration',
      subscribersDir: 'src/subscriber'
    }
  },
  test: {
    name: 'test',
    type: DB_TYPE,
    host: DB_HOST,
    port: DB_PORT,
    username: DB_USER,
    password: DB_PASS,
    database: DB_NAME_TEST,
    dropSchema: true,
    synchronize: true,
    logging: false,
    entities: ['src/entity/**/*.ts'],
    migrations: ['src/migration/**/*.ts'],
    subscribers: ['src/subscriber/**/*.ts'],
    cli: {
      entitiesDir: 'src/entity',
      migrationsDir: 'src/migration',
      subscribersDir: 'src/subscriber'
    }
  },
  production: {
    name: 'production',
    type: DB_TYPE,
    host: DB_HOST,
    port: DB_PORT,
    username: DB_USER,
    password: DB_PASS,
    database: DB_NAME_PRODUCTION,
    entities: ['src/entity/**/*.ts'],
    migrations: ['src/migration/**/*.ts'],
    subscribers: ['src/subscriber/**/*.ts'],
    cli: {
      entitiesDir: 'src/entity',
      migrationsDir: 'src/migration',
      subscribersDir: 'src/subscriber'
    }
  },
  root: {
    name: 'root',
    type: DB_TYPE,
    host: DB_HOST,
    port: DB_PORT,
    username: ROOT_USER,
    password: ROOT_PASS
  }
};

export const connectDb = async () => {
  const connectionOptions = c[process.env.NODE_ENV as string];

  return createConnection({
    ...connectionOptions,
    name: 'default'
  });
};

export const connectDbTest = async (resetDB: boolean = false) => {
  const connectionOptions = c[process.env.NODE_ENV as string];

  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB
  });
};
