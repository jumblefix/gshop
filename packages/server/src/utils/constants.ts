import { createError } from 'apollo-errors';

export const REDIS_SESSION_PREFIX = 'sess:';
export const USER_SESSION_PREFIX = 'userSids:';
export const FORGOT_PASSWORD_PREFIX = 'forgotPassword:';
export const ITEMS_PER_PAGE = 10;

export enum Env {
  development = 'development',
  test = 'test',
  production = 'production',
}

export enum TokenTypes {
  confirm = 'confirm',
  reset = 'reset',
}

export const ERROR_VALIDATION_FAILED = 'Validation failed';
export const ERROR_PASSWORDS_DONT_MATCH = 'Passwords do not match';
export const ERROR_INVALID_TOKEN = 'Token is Invalid or Expired';
export const ERROR_INVALID_CATEGORY = 'Invalid Category';
export const ERROR_ITEM_NOT_FOUND = 'Item not found';
export const ERROR_USER_NOT_FOUND = 'User not found';
export const ERROR_INVALID_LOGIN = 'Invalid email or password';
export const ERROR_ALREADY_IN_CART = 'Already in Cart';
export const ERROR_EMPTY = 'Empty';
export const ERROR_LOGIN_TO_CONTINUE = 'Please login to continue';
export const ERROR_PERMISSION_DENIED =
  'You dont have permission to perform this operation';

export const InputValidationError = createError('InputValidationError', {
  message: ERROR_VALIDATION_FAILED,
});

export const AuthenticationError = createError('AuthenticationError', {
  message: ERROR_LOGIN_TO_CONTINUE,
});

export const AuthorizationError = createError('AuthorizationError', {
  message: ERROR_PERMISSION_DENIED,
});
