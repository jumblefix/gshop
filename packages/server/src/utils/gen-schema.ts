import * as fs from 'fs';
import * as glob from 'glob';
import { makeExecutableSchema } from 'graphql-tools';
import { mergeResolvers, mergeTypes } from 'merge-graphql-schemas';
import * as path from 'path';

export function genSchema() {
  const pathToModules = path.join(__dirname, '../modules');

  const paths = glob.sync(`${pathToModules}/**/*.graphql`);
  const graphqlTypes = paths.map(x => fs.readFileSync(x, { encoding: 'utf8' }));
  const resolversList = glob
    .sync(`${pathToModules}/**/**.resolver.?s`)
    .map(resolver => require(resolver).resolvers);

  return makeExecutableSchema({
    typeDefs: mergeTypes(graphqlTypes),
    resolvers: mergeResolvers(resolversList),
  });
}
