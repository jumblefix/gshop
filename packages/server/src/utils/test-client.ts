import { gql } from 'apollo-server-express';
import { print } from 'graphql';
import MutationQueries from './test-mutation-queries';

export default class TestClient extends MutationQueries {
  url: string;
  testing: boolean;
  constructor(url: string, testing: boolean = false) {
    super(url, testing);
    this.url = url;
    this.testing = testing;
  }

  async me() {
    const query = print(gql`
      {
        me {
          id
          name
          email
        }
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async getCart() {
    const query = print(gql`
      {
        getCart {
          id
          product {
            id
            title
            price
          }
          user {
            id
            name
            email
          }
        }
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async listMainCategories() {
    const query = print(gql`
      {
        listMainCategories {
          id
          name
          slug
        }
      }
    `);

    return this.returnQueryOrPromise(query);
  }

  async getChildCategories(id: string) {
    const query = print(gql`
        {
          getChildCategories(id: "${id}") {
            id
            name
            slug
            children {
              id
              name
              slug
              children {
                id
                name
                slug
              }
            }
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async getBreadCrumbPath(id: string) {
    const query = print(gql`{
        getBreadCrumbPath(id: "${id}") {
          id
          name
          slug
          parent {
            id
            name
            slug
            parent {
              id
              name
              slug
            }
          }
        }
      }`);
    return this.returnQueryOrPromise(query);
  }

  async listProducts() {
    const query = print(gql`
      {
        listProducts {
          id
          title
          price
          slug
          description
        }
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async getProduct(id: string) {
    const query = print(gql`
        {
          getProduct(id: "${id}") {
            id
            title
            price
            slug
            description
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async getProductByCategory(id: string) {
    const query = print(gql`
        {
          getProductByCategory(categoryId: "${id}") {
            id
            title
            price
            slug
            description
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async getUser(id: string) {
    const query = print(gql`
        {
          getUser(id: "${id}") {
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async listUsers() {
    const query = print(gql`
      {
        listUsers {
          id
          name
          email
        }
      }
    `);
    return this.returnQueryOrPromise(query);
  }
}
