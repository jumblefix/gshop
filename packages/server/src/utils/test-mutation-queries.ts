import { gql } from 'apollo-server-express';
import Axios from 'axios';
import { print } from 'graphql';

export default class MutationQueries {
  url: string;
  testing: boolean;
  constructor(url: string, testing: boolean) {
    this.url = url;
    this.testing = testing;
  }

  returnQueryOrPromise(query: any) {
    if (this.testing) {
      return query;
    }
    return Axios.post(this.url, { query });
  }

  async register(name: string, email: string, password: string) {
    const query = print(gql`
        mutation {
          register(name: "${name}", email: "${email}", password: "${password}") {
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async addToCart(productId: string, quantity: number = 1) {
    const query = print(gql`
        mutation {
          addToCart(productId: "${productId}", quantity: ${quantity}) {
            id
            product {
              id
              title
              price
            }
            user {
              id
              name
              email
            }
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async removeFromCart(productId: string) {
    const query = print(gql`
        mutation {
          removeFromCart(productId:"${productId}")
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async emptyCart() {
    const query = print(gql`
      mutation {
        emptyCart
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async addCategory(name: string, parent: string) {
    const query = print(gql`
        mutation {
          addCategory(name: "${name}", parentId: "${parent}") {
            id
            name
            slug
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async addProduct(
    title: string,
    coverImage: string,
    description: string,
    rating: number,
    price: number,
    offerPrice: string,
    categoryId: string
  ) {
    const query = print(gql`
        mutation {
          addProduct(
            title: "${title}"
            coverImage:  "${coverImage}"
            description:  "${description}"
            rating: ${rating}
            price: ${price}
            offerPrice: ${offerPrice}
            categoryId: "${categoryId}"
          ) {
            id
            title
            slug
            price
            description
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async login(email: string, password: string) {
    const query = print(gql`
        mutation {
          login(email: "${email}", password: "${password}") {
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async logout() {
    const query = print(gql`
      mutation {
        logout
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async resendVerifySignup() {
    const query = print(gql`
      mutation {
        resendVerifySignup
      }
    `);
    return this.returnQueryOrPromise(query);
  }

  async sendResetPassword(email: string) {
    const query = print(gql`
        mutation {
          sendResetPassword(email:"${email}")
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async verifyResetPassword(
    token: string,
    password: string,
    confirmPassword: string
  ) {
    const query = print(gql`
        mutation {
          verifyResetPassword(token:"${token}",password:"${password}",confirmPassword:"${confirmPassword}"){
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async changePassword(oldPassword: string, password: string) {
    const query = print(gql`
        mutation {
          changePassword(oldPassword:"${oldPassword}",password:"${password}"){
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }

  async changeEmail(email: string) {
    const query = print(gql`
        mutation {
          changeEmail(email:"${email}"){
            id
            name
            email
          }
        }
      `);
    return this.returnQueryOrPromise(query);
  }
}
