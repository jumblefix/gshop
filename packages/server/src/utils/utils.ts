export const multiplyAll = (...args: number[]) =>
  args.reduce((c, p) => c * p, 1);
