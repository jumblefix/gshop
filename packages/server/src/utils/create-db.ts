import { createConnection } from 'typeorm';
import { c } from './connect-db';
import { Env } from './constants';

export async function createDb() {
  const connectionOptions = c.root;

  const rootConnection = await createConnection({
    ...connectionOptions,
    name: 'default'
  });

  let dbName: string;

  if (process.env.NODE_ENV === Env.test) {
    dbName = process.env.DB_NAME_TEST as string;
  } else if (process.env.NODE_ENV === Env.development) {
    dbName = process.env.DB_NAME as string;
  } else {
    dbName = process.env.DB_NAME_PRODUCTION as string;
  }

  const grantQ =
    // tslint:disable-next-line:prefer-template
    'GRANT ALL ON ' + dbName + '.* TO `' + process.env.DB_USER + '`@`%`;';

  await rootConnection
    .query(`CREATE DATABASE IF NOT EXISTS ${dbName};`)
    .then(async () => {
      await rootConnection.query(grantQ);
      await rootConnection.close();
    });
}
