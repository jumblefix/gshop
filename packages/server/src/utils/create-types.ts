import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs';
import * as path from 'path';
import { genSchema } from './gen-schema';

const typescriptTypes = generateNamespace('GQL', genSchema());

fs.writeFile(
  path.join(__dirname, '../typings/GQL/index.d.ts'),
  typescriptTypes,
  err => {
    console.error(err);
  },
);
