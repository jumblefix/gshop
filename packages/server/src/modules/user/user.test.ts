import { graphql } from 'graphql';
import { Connection } from 'typeorm';
import { connectDbTest } from '../../utils/connect-db';
import { createDb } from '../../utils/create-db';
import { genSchema } from '../../utils/gen-schema';
import TestClient from '../../utils/test-client';

const client = new TestClient('', true);

let connection: Connection;

beforeAll(async () => {
  try {
    await createDb();
  } catch (error) {
    throw new Error(error);
  }
  connection = await connectDbTest(true);
});

describe('user', () => {
  describe('me', () => {
    it('should return null if not logged in', async () => {
      const ctx = { session: {}, db: connection };
      const query: any = await client.me();
      const result = await graphql(genSchema(), query, null, ctx, {});
      expect(result).toMatchSnapshot();
    });
  });
});
