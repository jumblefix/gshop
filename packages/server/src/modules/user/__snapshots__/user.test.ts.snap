// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user me should return null if not logged in 1`] = `
Object {
  "data": Object {
    "me": null,
  },
}
`;
